### PR TITLE
feat: reload failed IPFS tabs when API becomes available

### PR DESCRIPTION
**Pull request by whizzzkid**
Closes https://github.com/ipfs/ipfs-companion/issues/1010

In this PR:

- I added support for reloading failed tabs when API is available again, by:
    - By statically checking for failed tabs which belongs to local gateway instead of tracking all URLs.
    - This reduces the complexity for tracking failed websites in `content_scripts`.
    - More details on the implementation can be found in the relevant reloader.
- I am also introducing a plugin mechanism that can register different reloader classes, this provides decent segregation of validation logic and also allows us to have reusable code for reloading.
- Added relevant unit tests for 100% coverage.
- Also updated the `yarn.lock` file and added a matching `.editorconfig` to configure IDEs.

Screencast:

https://user-images.githubusercontent.com/1895906/186346687-5de9caec-c4fc-4897-91cc-bed07990ba9d.mp4

### CODE REVIEW

File: .editorconfig
Comment (whizzzkid) [inline .editorconfig:8]: just to remove ambiguity from IDEs

File: add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js
Comment (whizzzkid) [inline add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js:22]: this logic validates offline pages belonging to IPFS node.

File: package.json
Comment (whizzzkid) [inline package.json:144]: for some reason latest LTS (18) fails to build this. I did not investigate, but added check to prevent this.

**Review by whizzzkid — COMMENTED** [review]
Adding explainer comments

File: add-on/src/lib/ipfs-client/index.js
Comment (SgtPooki) [inline add-on/src/lib/ipfs-client/index.js:72]: extensions should be async (if they're not already). So we should not use a forEach to loop over a reload operation.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#sect2 and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#sect3 for reasons why

Comment (SgtPooki) [reply add-on/src/lib/ipfs-client/index.js:72]: If we don't want to `await` each reload sequentially, that's fine too, but we should still wrap in a try/catch so we don't have floating promises.

File: add-on/src/lib/ipfs-client/index.js
Comment (SgtPooki) [inline add-on/src/lib/ipfs-client/index.js:86]: nit: I would prefer this be defined above

I would also prefer explicit args (`(a1, a2, a3) => reloadIpfsClientDependents(a1, a2, a3, [LocalGatewayReloader])`) since we don't have typescript to help with argument safety.

File: add-on/src/lib/ipfs-client/index.js
Comment (SgtPooki) [inline add-on/src/lib/ipfs-client/index.js:65]: add jsdoc comment

File: add-on/src/lib/ipfs-client/index.js
Comment (SgtPooki) [inline add-on/src/lib/ipfs-client/index.js:49]: try/catch for async function.

File: add-on/src/lib/ipfs-client/reloaders/index.js
Comment (SgtPooki) [inline add-on/src/lib/ipfs-client/reloaders/index.js:8]: Where is `ReloaderBase` defined? You should use jsdocs `@typedef` to define it. 

Same for Browser & Logger if they're not already defined (tip: You can use `{import('./some/path').Namespace['property']}` if the type is defined elsewhere)

File: add-on/src/lib/ipfs-client/reloaders/index.js
Comment (SgtPooki) [inline add-on/src/lib/ipfs-client/reloaders/index.js:18]: try/catch

File: add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js
Comment (SgtPooki) [inline add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js:22]: Use multiline comment with jsdoc tags where appropriate.

File: package.json
Comment (SgtPooki) [reply package.json:144]: +1. a lot of our packages and their dependencies are out of date, and there have been a lot of changes made w.r.t. esm support, so we've been in a tough spot. However, you can see our support goals at https://github.com/ipfs/ipfs-gui/tree/SgtPooki-version-support-definition#nodejs (We can link to main branch when https://github.com/ipfs/ipfs-gui/pull/97 gets merged)

**Review by SgtPooki — CHANGES_REQUESTED** [review]

File: add-on/src/lib/ipfs-client/index.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/index.js]: Added the try-catch block, it won't catch the promises inside the `forEach` block though, unless we explicitly await. I wanted this to be fire and forget.

File: add-on/src/lib/ipfs-client/index.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/index.js]: Done!

File: add-on/src/lib/ipfs-client/index.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/index.js]: Added the comment.

File: add-on/src/lib/ipfs-client/index.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/index.js]: I am not touching the existing functionality for now, try-catch would be redundant here, since we're not awaiting for the promise to resolve. So the error would never be caught.

File: add-on/src/lib/ipfs-client/reloaders/index.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/reloaders/index.js]: Fixed.

File: add-on/src/lib/ipfs-client/reloaders/index.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/reloaders/index.js]: Added.

File: add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js]: Fixed.

**Comment by whizzzkid** [comment]
@SgtPooki looks like the builds are failing on `windows-latest` I created a PR to address that: #1094 can you please try that?

**Review by lidel — CHANGES_REQUESTED** [review]
Thank you @whizzzkid, before we land this, do you mind 
- rebasing this to include fix from #1094  (which already landed in main branch)
- improving URL check to also work for subdomain gateways (details inline)
- (optional) perhaps we could avoid title matching by leveraging `onErrorOccurred`? (details inline – not a blocker, could be descoped and filled as an issue for future improvement)

File: add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js:25]: Thanks, @lidel

- the check should be `isIPFS.url || isIPFS.subdomain` and not `isIPFS.url && isIPFS.subdomain` added that.
- I'll create a separate issue for that, It was on my mind but this PR is large as it is, I'll work on that issue separately.

File: add-on/src/lib/ipfs-client/reloaders/webUiReloader.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/reloaders/webUiReloader.js:14]: This functionality was brought in from: https://github.com/ipfs/ipfs-companion/blob/main/add-on/src/lib/ipfs-client/index.js#L64-L68

But now that I think more about it, this will be handled in the `localGatewayReloader` check as `isIPFS.url || isIPFS.subdomain` returns true for `http://127.0.0.1:5001/ipfs/bafybeibozpulxtpv5nhfa2ue3dcjx23ndh3gwr5vwllk7ptoyfwnfjjr4q` hardcoded cid.

**Comment by whizzzkid** [comment]
@lidel this is ready for re-review:

- [x] rebasing this to include fix from https://github.com/ipfs/ipfs-companion/pull/1094 (which already landed in main branch)
- [x] improving URL check to also work for subdomain gateways (details inline)
(optional) perhaps we could avoid title matching by leveraging onErrorOccurred? (details inline – [x] not a blocker, could be descoped and filled as an issue for future improvement, lmk if you prefer that): #1097

Build is now 🟢 https://github.com/whizzzkid/ipfs-companion/actions/runs/3033417199

❤️

**Comment by BigLep** [comment]
@SgtPooki : is there any other feedback, or can this be merged?

**Comment by SgtPooki** [comment]
GUI triage: Lidel is going to give a look at this. We need to fix companion build pipeline before merging this

**Comment by whizzzkid** [comment]
@SgtPooki are the build pipelines broken again? I fixed the companion pipeline in https://github.com/ipfs/ipfs-companion/pull/1094

File: add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js
Comment (SgtPooki) [reply add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js:25]: @whizzzkid mind linking to the created issue or creating it and commenting here?

**Review by SgtPooki — APPROVED** [review]

**Comment by SgtPooki** [comment]
> @SgtPooki are the build pipelines broken again? I fixed the companion pipeline in #1094

Sorry we missed that =) Gave my approval but I've got very little experience in this repo so I'm leaving the PR open for lidel to handle when he gets to it. I'd ping him Monday if it's not merged by then

File: add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js
Comment (whizzzkid) [reply add-on/src/lib/ipfs-client/reloaders/localGatewayReloader.js:25]: Here, https://github.com/ipfs/ipfs-companion/issues/1097

Can you please assign this to me @SgtPooki/@lidel

**Comment by SgtPooki** [comment]
GUI triage: Lidel is going to give a look at this. We need to fix companion build pipeline before merging this

**Review by lidel — APPROVED** [review]
Thank you @whizzzkid and @SgtPooki 

LGTM, there are unrelated issues caused by old dependencies (will fill separate issues if I have no time to fix them myself), but the functionality of this PR is good!

 I'm merging it to unblock follow-up work from https://github.com/ipfs/ipfs-companion/issues/1097

**Pull request merged**
